core/src/main/java/org/apache/calcite/plan/RelOptUtil.java

```diff
@@ -3734,6 +3739,12 @@ private Exists(RelNode r, boolean indicator, boolean outerJoin) {
       this.outerJoin = outerJoin;
     }
   }
+
+  /** Check if it is the join whose condition is based on column equality,
+   * this mostly depends on the physical implementation of the join. */
+  public static boolean forceEquiJoin(Join join) {
+    return join.isSemiJoin() || join instanceof EquiJoin;
```

Comment:
Now you can remove `forceEquiJoin ` and just use `return join.analyzeCondition().isEqui();`

Reply:
No, we can not do this, because `join.analyzeCondition().isEqui()` <> `forceEquiJoin`, this method tell us if the join's condition must always keep as equi but not returns whether it is current a equi-join.

Reply:
That is a wrong assumption. SemiJoin's join condition doesn't need to be equi condition.

Reply:
I agree with @hsyuan , with the current PR we should start supporting non-equi semi joins. As defined in EnumerableJoinRule:
- Equi semi joins will run via EnumerableHashJoin (implementHashSemiJoin)
- Non-equi semi joins will run via EnumerableNestedLoopJoin

Reply:
Regarding my previous comment and in order to include this feature in 1.20, I guess we could leave the code as it is, and handle separately for 1.21 another ticket to manage non-equi semi joins.

Reply:
Thanks a lot @rubenada  and @hsyuan , I saw that @hsyuan had fired an issue to support non-equi semi join(also hash join and merge join), lets's solve the ticket in 1.21.

Reply:
@danny0405 Can you mark method `forceEquiJoin` subject to change or remove? I view it as a temporary walk around if I don't get it wrong.